docs/community/training/debugging/techniques-and-tools.md

```diff
@@ -0,0 +1,25 @@
+# Techniques and tools
+
+## Identifying errors
+
+How to read tracebacks?
+
+## Developing a plan
+
+Version control, last known good version, `git bisect`.
+
+## Searching for the root cause
+
+Print statements, debuggers, REPLs, last good version, recent changes, assert statements, comment out code ...
+
+## Fixing the root cause
+
+Is there an optimal solution?
+Smallest change or least effect on the rest of your code?
+Does the best solution involve restructuring or modifying other parts of your code?
+
+## After it's fixed
+
+Test cases, defensive programming, shorten the feedback loop, what coding practices might make such a mistake easier to find next time (e.g., small functions, frequent commits).
```

Comment:
I agree about test cases for after its fixed, but should we also suggest test cases for when its found? 
I personally think of tests that make sure the bug is reproduced before fixing code, then when I have fixed it, I have a suite of tests that should pass by ensuring the bad state is not hit. That way if it sneaks back in (which it has in the past) I catch it.

Reply:
I agree in principle, my only reservation is whether this would require a reasonably detailed treatment of writing (and running) test cases, and whether that's feasible to cover within the same workshop. But hey, maybe an example of test cases for the Python vs R example would be good, in that they'd be more than a toy example?